pysdk fs.py hardlink may return success but target is ENOENT after symlink loop scenario

**What happened**:

`pysdk` workflow 中 `Run fs.py` 步骤偶发失败。失败 job：

https://github.com/juicedata/juicefs/actions/runs/29772343861/job/88453386339

失败发生在 Hypothesis 状态机测试 `.github/scripts/hypo/fs.py` 的 `hardlink` 规则中：

```python
result1 = self.fsop1.do_hardlink(src_file=src_file, parent=parent, link_file_name=link_file_name, user=user, umask=umask)
result2 = self.fsop2.do_hardlink(src_file=src_file, parent=parent, link_file_name=link_file_name, user=user, umask=umask)
assert self.equal(result1, result2), red(f'hardlink:\nresult1 is {result1}\nresult2 is {result2}')
```

日志中的核心差异是：本地文件系统 hardlink 成功后，目标路径可以 `stat`；但 JuiceFS Python SDK / libjfs hardlink 返回成功后，目标路径随后 `stat` 却返回 `ENOENT`。

```text
do_hardlink /tmp/fsrand/aa/a/a /tmp/fsrand/a with user=root umask=0o22 succeed
do_stat /tmp/fsrand/a with user=root succeed

do_hardlink /tmp/jfs/fsrand/aa/a/a /tmp/jfs/fsrand/a with user=root umask=0o22 succeed
do_stat /tmp/jfs/fsrand/a {'entry': 'a', 'user': 'root'} failed: OSError:call jfs_stat failed: [Errno 2] No such file or directory: ["b'/fsrand/a'", "<cparam 'P' ...>"]
```

最终断言失败：

```text
AssertionError: hardlink:
result1 is (0, 0, '0o40755')
result2 is OSError:call jfs_stat failed: [Errno 2] No such file or directory: ["b'/fsrand/a'", "<cparam 'P' ...>"]
```

该失败前后日志还多次出现循环 symlink 相关信息：

```text
find a loop symlink
unknown errno 40: too many levels of symbolic links
```

因此问题看起来与 hardlink 过程中对 symlink / 循环 symlink 的路径解析或错误返回处理有关。当前行为可能是 `jfs_link` 在某些路径解析失败或目标创建失败的情况下仍返回成功，导致出现“hardlink 返回成功，但目标路径不存在”的不一致状态。

**What you expected to happen**:

JuiceFS Python SDK / libjfs 的 hardlink 行为应与本地文件系统一致：

- 如果 hardlink 返回成功，目标路径必须存在，并且随后 `stat(dst)` 应成功；
- 如果源路径解析失败，例如遇到循环 symlink，应返回对应 errno，例如 `ELOOP`；
- 如果目标无法创建，应返回对应错误；
- 不应出现 `jfs_link` 返回成功但 `jfs_stat(dst)` 返回 `ENOENT` 的情况。

**How to reproduce it (as minimally and precisely as possible)**:

该问题在 CI 中由 Hypothesis 随机状态机触发，因此在普通 CI run 中表现为概率性暴露；但日志里 Hypothesis 已经给出了可复现的最小失败样例。

1. 使用失败 job 对应代码版本：`605ff4c5ddb21337d0d9e8af6ed09544971a5bb4`。
2. 在 `.github/scripts/hypo/fs.py` 中给 `JuicefsMachine` 测试类临时加上 Hypothesis 提供的复现装饰器：

```python
from hypothesis import reproduce_failure

@reproduce_failure(
    '6.157.1',
    b'AXicY2BwZHRUZXAEwqbERAZHOSCjoTHRkQEsBIR8DCiAEagEogLIkmEAkQJghQyOCkCyOSW1DMQGm8jYmAgxQ5EBqJ4JYiYTyEImXCyoUUzk8eVB9gDNAgA0EhxF'
)
@seed(SEED)
class JuicefsMachine(RuleBasedStateMachine):
    ...
```

3. 按 `pysdk` workflow 的 `Run fs.py` 步骤运行测试：

```bash
sudo IGNORE_DIFF_ERRORS=true \
  MAX_EXAMPLE=1000 \
  STEP_COUNT=200 \
  USE_SDK=true \
  SEED=0 \
  META_URL=<meta_url> \
  EXCLUDE_RULES="readlines,readline,clone_cp_file,clone_cp_dir" \
  python3 .github/scripts/hypo/fs.py 2>&1 | tee fsrand.log
```

也可以直接参考失败 workflow：`.github/workflows/pysdk.yml` 中 `Run fs.py` 步骤。

**Anything else we need to know?**

推荐修复方向：

1. 检查 Python SDK / libjfs 中 `jfs_link` 对应的实现，以及底层 VFS / meta `Link` 路径。
2. 确保 hardlink 过程中对 source path 和 destination parent 的路径解析错误不会被吞掉或错误映射。
3. 如果解析 source path 时遇到循环 symlink，应返回 `ELOOP`，不要继续创建 link，也不要返回成功。
4. 如果底层 meta link 操作失败，应把错误原样返回给 Python SDK。
5. 可以增加防御性校验：`Link` 返回成功前确认目标 dentry 已经创建；若 lookup 目标失败，则返回对应错误而不是成功。

伪代码示例：

```go
func (v *VFS) Link(ctx Context, srcPath, dstPath string) syscall.Errno {
    srcParent, srcName, err := v.resolveParent(srcPath)
    if err != 0 {
        return err // includes ELOOP
    }

    dstParent, dstName, err := v.resolveParent(dstPath)
    if err != 0 {
        return err
    }

    srcIno, attr, err := v.lookup(srcParent, srcName)
    if err != 0 {
        return err
    }

    if attr.Typ == meta.TypeDirectory {
        return syscall.EPERM
    }

    err = v.Meta.Link(ctx, srcIno, dstParent, dstName, attr)
    if err != 0 {
        return err
    }

    // Defensive check: do not report success if destination does not exist.
    _, _, err = v.lookup(dstParent, dstName)
    if err != 0 {
        return err
    }

    return 0
}
```

不建议仅在 CI 中排除 `hardlink` 规则，例如把 `hardlink` 加入 `EXCLUDE_RULES`。这只能隐藏问题，不能修复 SDK 与本地文件系统语义不一致的问题。

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version: commit `605ff4c5ddb21337d0d9e8af6ed09544971a5bb4`, built in `pysdk` workflow
- Cloud provider or hardware configuration running JuiceFS: GitHub Actions runner `ubuntu-22.04`
- OS (e.g `cat /etc/os-release`): Ubuntu 22.04 GitHub-hosted runner
- Kernel (e.g. `uname -a`): GitHub-hosted runner kernel, see job environment
- Object storage (cloud provider and region, or self maintained): local bucket path configured by workflow, `--bucket=/mnt/jfs`
- Metadata engine info (version, cloud provider managed or self maintained): workflow matrix metadata engine; failed job log should be checked for exact matrix value
- Network connectivity (JuiceFS to metadata engine, JuiceFS to object storage): GitHub Actions CI environment
- Others: `USE_SDK=true`, `IGNORE_DIFF_ERRORS=true`, `MAX_EXAMPLE=1000`, `STEP_COUNT=200`, `EXCLUDE_RULES="readlines,readline,clone_cp_file,clone_cp_dir"`

--- a/pkg/fs/fs.go
+++ b/pkg/fs/fs.go
@@ -1031,12 +1031,18 @@ func (fs *FileSystem) doResolve(ctx meta.Context, p string, followLastSymlink bo
 			} else {
 				target = path.Join(strings.Join(ss[:i], "/"), target)
 			}
-			fi, err = fs.doResolve(ctx, target, followLastSymlink, visited)
-			if err != 0 {
-				return
+			hasRemaining := false
+			for _, comp := range ss[i+1:] {
+				if comp != "" {
+					target = target + "/" + comp
+					hasRemaining = true
+				}
 			}
-			inode = fi.Inode()
-			attr = fi.attr
+			fi, err = fs.doResolve(ctx, target, followLastSymlink, visited)
+			if err == 0 && fi != nil && !hasRemaining {
+				fi.name = name
+			}
+			return
 		}
 		fi.name = name
 		parent = inode

--- a/pkg/fs/fs.go
+++ b/pkg/fs/fs.go
@@ -986,17 +986,14 @@ func (fs *FileSystem) doResolve(ctx meta.Context, p string, followLastSymlink bo
 		}
 
 		var inode Ino
-		var resolved bool
+		isLastComponent := i == len(ss)-1
 
 		err = fs.lookup(ctx, parent, name, &inode, attr)
-		if i == len(ss)-1 {
-			resolved = true
-		}
 		if err != 0 {
 			return
 		}
 		fi = AttrToFileInfo(inode, attr)
-		if (!resolved || followLastSymlink) && fi.IsSymlink() {
+		if (!isLastComponent || followLastSymlink) && fi.IsSymlink() {
 			if _, ok := visited[inode]; ok {
 				logger.Errorf("find a loop symlink: %d", inode)
 				return nil, syscall.ELOOP
@@ -1031,15 +1028,11 @@ func (fs *FileSystem) doResolve(ctx meta.Context, p string, followLastSymlink bo
 			} else {
 				target = path.Join(strings.Join(ss[:i], "/"), target)
 			}
-			hasRemaining := false
-			for _, comp := range ss[i+1:] {
-				if comp != "" {
-					target = target + "/" + comp
-					hasRemaining = true
-				}
+			if !isLastComponent {
+				target += "/" + strings.Join(ss[i+1:], "/")
 			}
 			fi, err = fs.doResolve(ctx, target, followLastSymlink, visited)
-			if err == 0 && fi != nil && !hasRemaining {
+			if err == 0 && isLastComponent {
 				fi.name = name
 			}
 			return

--- a/pkg/fs/fs_test.go
+++ b/pkg/fs/fs_test.go
@@ -344,6 +344,41 @@ func TestBatchDeleteEntriesChecksParentWriteAndSearchPermission(t *testing.T) {
 	}
 }
 
+func TestResolveRelativeSymlinkAfterRedirection(t *testing.T) {
+	fs := createTestFS(t)
+	ctx := meta.NewContext(1, 1, []uint32{2})
+
+	for _, dir := range []string{"/a", "/x/y", "/x/z"} {
+		if err := fs.MkdirAll(ctx, dir, 0777, 022); err != 0 {
+			t.Fatalf("mkdir %s: %s", dir, err)
+		}
+	}
+
+	f, err := fs.Create(ctx, "/x/z/file", 0666, 022)
+	if err != 0 {
+		t.Fatalf("create target file: %s", err)
+	}
+	if err := f.Close(ctx); err != 0 {
+		t.Fatalf("close target file: %s", err)
+	}
+
+	if err := fs.Symlink(ctx, "../x/y", "/a/link1"); err != 0 {
+		t.Fatalf("symlink /a/link1: %s", err)
+	}
+	if err := fs.Symlink(ctx, "../z", "/x/y/link2"); err != 0 {
+		t.Fatalf("symlink /x/y/link2: %s", err)
+	}
+
+	if fi, err := fs.Stat(ctx, "/a/link1/link2/file"); err != 0 {
+		t.Fatalf("stat nested relative symlink path: %s", err)
+	} else if fi.name != "file" {
+		t.Fatalf("unexpected final name: %+v", fi)
+	}
+	if err := fs.Close(); err != nil {
+		t.Fatalf("close: %s", err)
+	}
+}
+
 func createTestFS(t testing.TB) *FileSystem {
 	m := meta.NewClient("memkv://", nil)
 	format := &meta.Format{

--- a/sdk/java/libjfs/main.go
+++ b/sdk/java/libjfs/main.go
@@ -133,6 +133,7 @@ const (
 	ENOTEMPTY = -0x27
 	ENODATA   = -0x3d
 	ENOTSUP   = -0x5f
+	ELOOP     = -0x28
 )
 
 func errno(err error) int32 {
@@ -178,6 +179,8 @@ func errno(err error) int32 {
 		return ENODATA
 	case syscall.ENOTSUP:
 		return ENOTSUP
+	case syscall.ELOOP:
+		return ELOOP
 	default:
 		logger.Warnf("unknown errno %d: %s", eno, err)
 		return -int32(eno)

